File: web/dashboard/src/types/delivery.ts
import type { ChatConversation, ChatMessage } from './chat'
import type { NotificationDLQ, NotificationTrace, OrderTimeline } from './notification'
import type { Order } from './order'

export type DeliveryRecordKind = 'notification' | 'chat' | 'ack' | 'dlq' | 'retry'

export interface DeliveryMessage {
  id: string
  message_id: string
  record_kind: DeliveryRecordKind
  type: string
  order_id?: string
  sender?: string
  target?: string
  delivery_method?: string
  status: string
  priority?: string
  ttl_seconds?: number
  ack_policy?: string
  business_type?: string
  event_type?: string
  trace_id?: string
  retry_count?: number
  last_error?: string
  conversation_id?: string
  room_id?: string
  sender_uid?: number
  receiver_uid?: number
  sender_role?: string
  target_node?: string
  latency_ms?: number
  created_at: string
  updated_at?: string
}

export interface DeliveryAttemptView {
  attempt_id: string
  status: string
  path?: string
  target?: string
  target_node?: string
  error_message?: string
  latency_ms?: number
  started_at: string
  finished_at?: string
}

export interface DeliveryAckView {
  ack_id: string
  user_id: string
  msg_id?: string
  device_id?: string
  ack_key?: string
  latency_ms?: number
  created_at: string
}

export interface DeliveryMessageDetail {
  message: DeliveryMessage
  raw_payload?: unknown
  business_type?: string
  event_type?: string
  trace_id?: string
  delivery_path?: string
  attempts?: DeliveryAttemptView[]
  acks?: DeliveryAckView[]
  retry_count?: number
  dlq?: NotificationDLQ
  failure_reason?: string
  last_error?: string
  target_node?: string
  latency_ms?: number
  order?: Order
  notification_trace?: NotificationTrace
  order_timeline?: OrderTimeline
  chat_message?: ChatMessage
  conversation?: ChatConversation
}

File: internal/notify/model/delivery.go
package model

import "time"

// DeliveryMessage is a flattened row for the dashboard delivery status page.
type DeliveryMessage struct {
	ID             string    `json:"id"`
	MessageID      string    `json:"message_id"`
	RecordKind     string    `json:"record_kind"`
	Type           string    `json:"type"`
	OrderID        string    `json:"order_id,omitempty"`
	Sender         string    `json:"sender,omitempty"`
	Target         string    `json:"target,omitempty"`
	DeliveryMethod string    `json:"delivery_method,omitempty"`
	Status         string    `json:"status"`
	Priority       string    `json:"priority,omitempty"`
	TTLSeconds     int64     `json:"ttl_seconds,omitempty"`
	AckPolicy      string    `json:"ack_policy,omitempty"`
	BusinessType   string    `json:"business_type,omitempty"`
	EventType      string    `json:"event_type,omitempty"`
	TraceID        string    `json:"trace_id,omitempty"`
	RetryCount     int64     `json:"retry_count,omitempty"`
	LastError      string    `json:"last_error,omitempty"`
	ConversationID string    `json:"conversation_id,omitempty"`
	RoomID         string    `json:"room_id,omitempty"`
	SenderUID      int64     `json:"sender_uid,omitempty"`
	ReceiverUID    int64     `json:"receiver_uid,omitempty"`
	SenderRole     string    `json:"sender_role,omitempty"`
	TargetNode     string    `json:"target_node,omitempty"`
	LatencyMs      float64   `json:"latency_ms,omitempty"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at,omitempty"`
}

// DeliveryMessageDetail expands one delivery row with raw trace material.
type DeliveryMessageDetail struct {
	Message           *DeliveryMessage       `json:"message"`
	RawPayload        interface{}            `json:"raw_payload,omitempty"`
	BusinessType      string                 `json:"business_type,omitempty"`
	EventType         string                 `json:"event_type,omitempty"`
	TraceID           string                 `json:"trace_id,omitempty"`
	DeliveryPath      string                 `json:"delivery_path,omitempty"`
	Attempts          []*NotificationAttempt `json:"attempts,omitempty"`
	ACKs              []*NotificationAck     `json:"acks,omitempty"`
	RetryCount        int64                  `json:"retry_count,omitempty"`
	DLQ               *NotificationDLQ       `json:"dlq,omitempty"`
	FailureReason     string                 `json:"failure_reason,omitempty"`
	LastError         string                 `json:"last_error,omitempty"`
	TargetNode        string                 `json:"target_node,omitempty"`
	LatencyMs         float64                `json:"latency_ms,omitempty"`
	Order             *Order                 `json:"order,omitempty"`
	NotificationTrace *NotificationTrace     `json:"notification_trace,omitempty"`
	OrderTimeline     *OrderTimeline         `json:"order_timeline,omitempty"`
	ChatMessage       *ChatMessage           `json:"chat_message,omitempty"`
	Conversation      *ChatConversation      `json:"conversation,omitempty"`
}
